### PR TITLE
One Shot Modifier callbacks

### PR DESCRIPTION
**Pull request by thomasbaart**
## Description

I'd like for it to be easier to add custom logic to changes in One Shot modifiers. I added eight callbacks to the code, of which four apply to the `_user` and four to `_kb`, in order to let both a keyboard maker and its users customize functionality for One Shot modifiers.

As a user, you will be able to use these four callbacks:

```c
void oneshot_locked_mods_set_user(uint8_t mods) { println("User: Oneshot locked mods set"); }
void oneshot_locked_mods_cleared_user(void) { println("User: Oneshot locked mods cleared"); }
void oneshot_mods_set_user(uint8_t mods) { println("User: Oneshot mods set"); }
void oneshot_mods_cleared_user(void) { println("User: Oneshot mods cleared"); }
```

What I should still figure out:

- [ ] I couldn't get my unit test to compile yet. I have not made further tests.
- [ ] I should check for size impact on existing keymaps.
- [x] I should document the order of callbacks better in the docs (see below).

Please let me know if you have any feedback, thoughts or comments. Do you think it will be useful to add this to the core? Is there a better way to handle this kind of functionality?

### Scenarios

I have tested the following three scenarios. The output is copied from the QMK Toolbox console. Each `>` denotes a matrix scan (output by QMK, not by me).

#### 1: Tap once, timeout ✅

Tap a One Shot Modifier once and let it time out. This behaves like I would expect.

```
> User: Oneshot mods set
> User: Oneshot mods cleared
```

#### 2: Tap twice ✅

Tap a One Shot Modifier twice. This behaves like I would expect.

```
> User: Oneshot mods set
> User: Oneshot mods cleared
```

#### 3: Lock it by tapping five times, then unlocking ❓

Tap the One Shot Modifier five times, and then unlock it with a sixth press. It behaves as expected up until the lock.

When unlocking, it does set the mods, then clear the locked mods and clear the oneshot mods, in that order. I could go two ways with this:

1. Apply some kind of logic to only call the "locked mods cleared" callback;
2. Leave it as is.

I'm leaning towards the second option, simply because it is what the framework does, and I don't know all future use cases users may come up with. Leaving it as is is the truth, and will not lead to unexpected behaviour in the future.

```
Tap one (set):
> User: Oneshot mods set

Tap two (clear):
> User: Oneshot mods cleared

Tap five (lock):
> User: Oneshot locked mods set

Tap six (unlock):
> User: Oneshot mods set
  User: Oneshot locked mods cleared
  User: Oneshot mods cleared
```

## Types of changes

- [x] Core
- [ ] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by this PR

* None

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tmk_core/common/action_util.c [hunk 1]
```diff
@@ -51,8 +51,18 @@ extern inline void clear_keys(void);
...
 int8_t get_oneshot_locked_mods(void) { return oneshot_locked_mods; }
-void set_oneshot_locked_mods(int8_t mods) { oneshot_locked_mods = mods; }
-void clear_oneshot_locked_mods(void) { oneshot_locked_mods = 0; }
+void set_oneshot_locked_mods(int8_t mods) {
+    if (mods != oneshot_locked_mods) {
+        oneshot_locked_mods = mods;
+        oneshot_locked_mods_set_kb(mods);
+    }
+}
+void clear_oneshot_locked_mods(int8_t mods) {
```
Comment (drashna) [inline tmk_core/common/action_util.c]: If this has been changed, then this needs to be changed as well: 
https://github.com/qmk/qmk_firmware/blob/f4012ba9b9abc357847cc4a3d0e696b933687a85/tmk_core/common/action_util.h#L76

Comment (drashna) [reply tmk_core/common/action_util.c]: Also, these probably should be `uint8_t`

```suggestion
void clear_oneshot_locked_mods(uint8_tmods) {
```

File: tmk_core/common/action_util.c [hunk 2]
```diff
@@ -243,23 +253,27 @@ void clear_macro_mods(void) { macro_mods = 0; }
...
+  }
 }
 /** \brief clear oneshot mods
  *
  * FIXME: needs doc
  */
-void clear_oneshot_mods(void)
-{
+void clear_oneshot_mods(void) {
+  if (oneshot_mods!= mods) {
```
Comment (drashna) [inline tmk_core/common/action_util.c]: `mods` is not defined.

This should probably be:
```suggestion
  if (oneshot_mods) {
```

As it's checking if `oneshot_mods` is actually set, or not.  If it's not set...

**Comment by drashna** [comment]
Have you had a chance to update based on the feedback here?

**Comment by thomasbaart** [comment]
Can't be accepted yet:

- [ ] Want to add layer callbacks too, to keep all oneshot functionality consistent. Easier to explain in docs, easier to understand for users
- [ ] Need to test the layer callback code, too
- [ ] Need to amend the docs

**Comment by drashna** [comment]
Sounds like a plan!

File: tmk_core/common/action_util.c [hunk 3]
```diff
@@ -130,10 +130,7 @@ void clear_oneshot_layer_state(oneshot_fullfillment_t state)
     oneshot_layer_data &= ~state;
     if (!get_oneshot_layer_state() && start_state != oneshot_layer_data) {
         layer_off(get_oneshot_layer());
-#if (defined(ONESHOT_TIMEOUT) && (ONESHOT_TIMEOUT > 0))
-        oneshot_layer_time = 0;
-#endif
-        oneshot_layers_changed_kb(get_oneshot_layer());
+        reset_oneshot_layer();
```
Comment (thomasbaart) [inline tmk_core/common/action_util.c]: Take special note of this change!

When clearing layer state is reset, the layer is switched off too. This change isn't reflected in the oneshot_layer_data variable, causing my callback to not reflect the current state. I think this is a bug, and so changed the method to call `reset_oneshot_layer` instead.

**Comment by thomasbaart** [comment]
I've added docs now, added layer change callbacks and tested the functionality. It's A-okay from my side now, barring any comments.

**Comment by drashna** [comment]
Awesome, it looks good!

**Pull request merged**